packages/ember-routing-views/lib/components/link-to.js

```diff
@@ -618,22 +618,54 @@ var LinkComponent = EmberComponent.extend({
       return false;
     }
 
-    var targetAttribute2 = this.attrs.target;
+    let targetAttribute2 = this.attrs.target;
     if (targetAttribute2 && targetAttribute2 !== '_self') {
       return false;
     }
 
-    var routing = get(this, '_routing');
-    var targetRouteName = this._computeRouteNameWithQueryParams(get(this, 'targetRouteName'));
-    var models = get(this, 'models');
-    var queryParamValues = get(this, 'queryParams.values');
-    var shouldReplace = get(this, 'attrs.replace');
+    let routing = get(this, '_routing');
+    let qualifiedRouteName = get(this, 'qualifiedRouteName');
+    let models = get(this, 'models');
+    let queryParamValues = get(this, 'queryParams.values');
+    let shouldReplace = get(this, 'attrs.replace');
 
-    routing.transitionTo(targetRouteName, models, queryParamValues, shouldReplace);
+    routing.transitionTo(qualifiedRouteName, models, queryParamValues, shouldReplace);
   },
 
   queryParams: null,
 
+  qualifiedRouteName: computed('targetRouteName', '_routing.currentState', function computeLinkToComponentQualifiedRouteName() {
+    let targetRouteName = get(this, 'targetRouteName');
+    let params = this.attrs.params.slice();
+    let lastParam = params[params.length - 1];
+    if (lastParam && lastParam.isQueryParams) {
+      params.pop();
+    }
+    let onlyQueryParamsSupplied = (this[HAS_BLOCK] ? params.length === 0 : params.length === 1);
+    if (onlyQueryParamsSupplied) {
+      let appController = this.container.lookup('controller:application');
+      if (appController) {
+        return get(appController, 'currentRouteName');
```

Comment:
This is available at `_routing.currentRouteName`, so no need to lookup `controller:application` (hopefully makes our future selves happy as we remove controllers)...